src/main/groovy/io/fourfinanceit/pipeline/example/jobs/BuildAndPublish.groovy

```diff
@@ -22,29 +29,22 @@ import static io.fourfinanceit.pipeline.example.JenkinsVariable.SCM_POLLING_ENAB
  * @author Marek Kapowicki
  */
 class BuildAndPublish extends MicroserviceJobDefinition {
+    
     @Override
     void configure(Job job, MicroserviceProject project, JenkinsVariables jenkinsVariables) {
-        job.with {
-            setupEnvironmentVariables(job)
-            configure JobDslConfigureHacksUtil.injectAndMaskPasswords()
-            scm MicroserviceJobUtils.configureGit(project)
-            if (project.scmConfig.cronToPoll && jenkinsVariables.getBoolean(SCM_POLLING_ENABLED, true)) {
-                triggers {
-                    scm(project.scmConfig.cronToPoll)
-                }
-            }
-            addGradleBuildStep(job, project)
-            //addArchiveJunitAsBuildStep(job)
-            steps {
-                ifBuildStillSuccessful(delegate as StepContext) {
-                    gradle("publish --stacktrace ${projectVersionParameters()}")
-                }
-            }
-        }
+        (
+            setupEnvironmentVariables() &
+                maskPasswords() &
+                configureGit(project) &
+                use(onCron(project.scmConfig.cronToPoll))
+                    .when(scmPoolingEnabled(jenkinsVariables)) &
+                buildProject() &
+                publishArtifacts()
+        ).applyOn(job)
```

Comment:
I still believe asking the user to remember/know upfront (!) that they have to wrap the composition into parentheses to use it correctly is wrong. Won't insist though, The Company (lawsuits! 😨) is the main user.

I don't get what you mean in 4.

Like the idea in 5., especially if it'll allow for different composition order (type-preserving in addition to type-narrowing). E.g.

```
job << abstractComponent | //type-preserving
  concreteComponent + //type-narrowing
  abstractComponent +
  concreteComponent
```